linting/extra/src/non_fallible_api.rs

```diff
@@ -0,0 +1,305 @@
+// Copyright (C) Parity Technologies (UK) Ltd.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use crate::ink_utils::{
+    expand_unnamed_consts,
+    find_contract_impl_id,
+};
+use clippy_utils::{
+    diagnostics::span_lint_and_then,
+    is_lint_allowed,
+    match_def_path,
+};
+use if_chain::if_chain;
+use rustc_errors::Applicability;
+use rustc_hir::{
+    self as hir,
+    def_id::DefId,
+    intravisit::{
+        walk_body,
+        walk_expr,
+        Visitor,
+    },
+    Body,
+    Expr,
+    ExprKind,
+    ImplItemKind,
+    ItemKind,
+    PathSegment,
+};
+use rustc_lint::{
+    LateContext,
+    LateLintPass,
+};
+use rustc_middle::{
+    hir::nested_filter,
+    ty::{
+        self,
+        ConstKind,
+        Ty,
+        TypeckResults,
+    },
+};
+use rustc_session::{
+    declare_lint,
+    declare_lint_pass,
+};
+use rustc_type_ir::sty::TyKind;
+
+declare_lint! {
+    /// ## What it does
+    ///
+    /// The lint detects potentially unsafe uses of methods for which there are safer alternatives.
+    ///
+    /// ## Why is this bad?
+    ///
+    /// In some standard collections in ink!, there are two types of implementations: non-fallible
+    /// (e.g. `get`) and fallible (e.g. `try_get`). Fallible alternatives are considered safer,
+    /// as they perform additional checks for incorrect input parameters and return `Result::Err`
+    /// when they are used improperly. On the other hand, non-fallible methods do not provide these
+    /// checks and will panic on incorrect input, placing the responsibility on the user to
+    /// implement these checks.
+    ///
+    /// For more context, see: [ink#1910](https://github.com/paritytech/ink/pull/1910).
+    ///
+    /// ## Example
+    ///
+    /// Consider the contract that has the following `Mapping` field:
+    ///
+    /// ```rust
+    /// #[ink(storage)]
+    /// pub struct Example { map: Mapping<String, AccountId> }
+    /// ```
+    ///
+    /// The following usage of the non-fallible API is unsafe:
+    ///
+    /// ```rust
+    /// pub fn test(&mut self, a: String, b: AccountId) {
```

Comment:
I think this exact case can actually not panic.

The input for this looks like this.
`[u8; 4]` + `String` + `[u8; 32]` = 36 plus the length of the encoded String (first 4 bytes are the message selector)

The `map.insert()` will encode:
`mapping base key` + `String` + `[u8; 32]`)  = 36 plus the length of the encoded String (the base key is 4 bytes)

Thus, if the message input can be decoded using the static buffer, that the encoded data is guaranteed to fit into the static buffer. IOW, any input to this message that would fail the storage operation will revert prior to that anyways (during the decoding of the input), rendering this usage of the fallible API unnecessary (and adding overhead).

To make the example valid you can for example push to the supplied string to grow it beyond the capacity of the static buffer. Or you could remove the `b: AccountID` function argument and just use the `AccountId` of the caller.

Reply:
That's a good point. Fixed.